File: services/core/holding_valuation_test.go
package main

import (
	"context"
	"net/http"
	"strings"
	"testing"
	"time"
)

const holdingValuationAsOf = "2026-01-11T14:00:00Z"

func TestHoldingValuationUsesExactNativePriceAndKeepsSnapshotUnavailable(t *testing.T) {
	svc := holdingValuationService(t)
	recordHoldingValuationPrice(t, svc, "fixture_aapl_old", "XNAS", "12", "2026-01-10T13:00:00Z", "2026-01-10T13:00:01Z", "2026-01-11T12:00:00Z")
	price := recordHoldingValuationPrice(t, svc, "fixture_aapl_xnas", "XNAS", "15", "2026-01-10T14:00:00Z", "2026-01-10T14:00:01Z", "2026-01-11T13:00:00Z")
	svc.now = func() time.Time { return mustTime(holdingValuationAsOf) }

	valuation, appErr := svc.holdingValuation(context.Background(), holdingValuationAsOf)
	if appErr != nil {
		t.Fatal(appErr)
	}
	if valuation.Scope != "holdings_only" || valuation.PolicyVersion != "native_holding_valuation_v1" ||
		valuation.MaxObservationAgeSeconds != 86400 || valuation.ValuationAsOf != holdingValuationAsOf ||
		valuation.LedgerRevision != "rev_0000000001" || valuation.Status != "stale_sample" || !valuation.Sample {
		t.Fatalf("holding valuation contract drifted: %+v", valuation)
	}
	if len(valuation.Lines) != 1 {
		t.Fatalf("expected one holding line: %+v", valuation.Lines)
	}
	line := valuation.Lines[0]
	if line.InstrumentID != "instrument_aapl" || line.Symbol != "AAPL" || line.Quantity != "3" ||
		line.CostBasis != "31" || line.Currency != "USD" || line.Status != "valued" || line.Issue != nil ||
		line.MarketValue == nil || *line.MarketValue != "45" || line.UnrealizedPnL == nil || *line.UnrealizedPnL != "14" {
		t.Fatalf("native holding math drifted: %+v", line)
	}
	if line.Price == nil || line.Price.ObservationID != price.ObservationID || line.Price.Source != "local_fixture" ||
		line.Price.Venue != "XNAS" || line.Price.Currency != "USD" || line.Price.Price != "15" ||
		line.Price.PriceAdjustment != marketDataAdjustmentUnspecified || line.Price.ObservedAt != "2026-01-10T14:00:00Z" ||
		line.Price.FetchedAt != "2026-01-10T14:00:01Z" || line.Price.RecordedAt != "2026-01-11T13:00:00Z" ||
		!line.Price.Sample || line.Price.State != "stale" {
		t.Fatalf("price provenance drifted: %+v", line.Price)
	}
	if len(valuation.Totals) != 1 || valuation.Totals[0].Currency != "USD" || valuation.Totals[0].CostBasis != "31" ||
		valuation.Totals[0].MarketValue != "45" || valuation.Totals[0].UnrealizedPnL != "14" {
		t.Fatalf("native currency totals drifted: %+v", valuation.Totals)
	}
	if len(valuation.Issues) != 1 || valuation.Issues[0].Code != "sample_data" {
		t.Fatalf("sample status is not explicit: %+v", valuation.Issues)
	}
	snapshot, err := snapshotFrom(context.Background(), svc.db)
	if err != nil || snapshot.ValuationStatus != "unavailable" {
		t.Fatalf("internal holding valuation changed public authority: snapshot=%+v err=%v", snapshot, err)
	}
}

func TestHoldingValuationSuppressesTotalsAndFailsClosed(t *testing.T) {
	tests := []struct {
		name     string
		prices   []struct{ id, venue, observed, fetched, recorded string }
		wantCode string
	}{
		{name: "missing", wantCode: "missing_security_price"},
		{name: "ambiguous venue", prices: []struct{ id, venue, observed, fetched, recorded string }{
			{"fixture_xnas", "XNAS", "2026-01-11T13:00:00Z", "2026-01-11T13:00:01Z", "2026-01-11T13:00:02Z"},
			{"fixture_bats", "BATS", "2026-01-11T13:01:00Z", "2026-01-11T13:01:01Z", "2026-01-11T13:01:02Z"},
		}, wantCode: "ambiguous_security_price_venue"},
		{name: "older than 24h boundary", prices: []struct{ id, venue, observed, fetched, recorded string }{
			{"fixture_stale", "XNAS", "2026-01-10T13:59:59.999999999Z", "2026-01-10T14:00:00Z", "2026-01-11T13:00:00Z"},
		}, wantCode: "stale_security_price"},
		{name: "future recorded is missing", prices: []struct{ id, venue, observed, fetched, recorded string }{
			{"fixture_future_recorded", "XNAS", "2026-01-11T13:00:00Z", "2026-01-11T13:00:01Z", "2026-01-11T14:00:01Z"},
		}, wantCode: "missing_security_price"},
		{name: "future observed is missing", prices: []struct{ id, venue, observed, fetched, recorded string }{
			{"fixture_future_observed", "XNAS", "2026-01-11T14:00:01Z", "2026-01-11T14:00:01Z", "2026-01-11T14:00:02Z"},
		}, wantCode: "missing_security_price"},
		{name: "future fetched is missing", prices: []struct{ id, venue, observed, fetched, recorded string }{
			{"fixture_future_fetched", "XNAS", "2026-01-11T13:00:00Z", "2026-01-11T14:00:01Z", "2026-01-11T14:00:02Z"},
		}, wantCode: "missing_security_price"},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			svc := holdingValuationService(t)
			for _, price := range test.prices {
				recordHoldingValuationPrice(t, svc, price.id, price.venue, "15", price.observed, price.fetched, price.recorded)
			}
			svc.now = func() time.Time { return mustTime("2026-01-11T15:00:00Z") }
			valuation, appErr := svc.holdingValuation(context.Background(), holdingValuationAsOf)
			if appErr != nil {
				t.Fatal(appErr)
			}
			if valuation.Status != "unavailable" || valuation.Totals != nil || len(valuation.Lines) != 1 ||
				valuation.Lines[0].Issue == nil || valuation.Lines[0].Issue.Code != test.wantCode ||
				len(valuation.Issues) != 1 || valuation.Issues[0].Code != test.wantCode {
				t.Fatalf("unavailable holding was aggregated or mislabeled: %+v", valuation)
			}
			for _, forbidden := range []string{"fixture_", "record_sha256", "SELECT ", "security_price_observations"} {
				if strings.Contains(valuation.Lines[0].Issue.Message, forbidden) || strings.Contains(valuation.Issues[0].Message, forbidden) {
					t.Fatalf("issue leaked internal detail %q: %+v", forbidden, valuation.Issues)
				}
			}
		})
	}

	t.Run("corrupt ledger revision", func(t *testing.T) {
		svc := holdingValuationService(t)
		if _, err := svc.db.Exec(`UPDATE ledger_meta SET revision=99 WHERE singleton=1`); err != nil {
			t.Fatal(err)
		}
		svc.now = func() time.Time { return mustTime(holdingValuationAsOf) }
		valuation, appErr := svc.holdingValuation(context.Background(), holdingValuationAsOf)
		if valuation != nil || appErr == nil || appErr.status != http.StatusInternalServerError || appErr.body.Message != "internal server error" {
			t.Fatalf("corrupt ledger revision was certified: valuation=%+v err=%v", valuation, appErr)
		}
	})
}

func holdingValuationService(t *testing.T) *Service {
	t.Helper()
	svc, _ := testService(t, nil, nil)
	applyTestCSV(t, svc, "holding-valuation-seed", testCSV(
		"aapl-buy,account-main,BUY,2026-01-10T10:00:00Z,AAPL,3,10,1,USD,-31",
	))
	return svc
}

func recordHoldingValuationPrice(t *testing.T, svc *Service, sourceID, venue, price, observedAt, fetchedAt, recordedAt string) *SecurityPriceObservation {
	t.Helper()
	svc.now = func() time.Time { return mustTime(recordedAt) }
	observation, err := svc.recordSecurityPriceObservation(context.Background(), SecurityPriceObservationInput{
		Source: "local_fixture", SourceObservationID: sourceID, InstrumentID: "instrument_aapl", Symbol: "AAPL",
		Venue: venue, Currency: "USD", Price: price, PriceAdjustment: marketDataAdjustmentUnspecified,
		ObservedAt: observedAt, FetchedAt: fetchedAt,
	})
	if err != nil {
		t.Fatal(err)
	}
	return observation
}

File: apps/client/test/holding_valuation_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:omni_folio_client/api.dart';
import 'package:omni_folio_client/app.dart';
import 'package:omni_folio_client/holding_valuation_page.dart';
import 'package:omni_folio_client/models.dart';
import 'package:omni_folio_client/paper_monitor.dart';

Json holdingValuationJson({
  String status = 'stale_sample',
  bool sample = true,
  List<Json>? totals,
  List<Json>? lines,
  List<Json>? issues,
}) => {
  'scope': 'holdings_only',
  'policy_version': 'native_holding_valuation_v1',
  'max_observation_age_seconds': 86400,
  'valuation_as_of': '2026-01-11T14:00:00Z',
  'ledger_revision': 'rev_0000000001',
  'ledger_as_of': '2026-01-10T10:00:00Z',
  'ledger_recorded_at': '2026-01-11T13:00:00Z',
  'status': status,
  'sample': sample,
  'totals':
      totals ??
      (status == 'stale_sample'
          ? const [
              {
                'currency': 'USD',
                'cost_basis': '31',
                'market_value': '45',
                'unrealized_pnl': '14',
              },
            ]
          : null),
  'lines':
      lines ??
      (status == 'empty'
          ? const []
          : const [
              {
                'symbol': 'AAPL',
                'quantity': '3',
                'cost_basis': '31',
                'currency': 'USD',
                'status': 'valued',
                'price': {
                  'source': 'local_fixture',
                  'venue': 'XNAS',
                  'currency': 'USD',
                  'price': '15',
                  'price_adjustment': 'unspecified',
                  'observed_at': '2026-01-10T14:00:00Z',
                  'fetched_at': '2026-01-10T14:00:01Z',
                  'recorded_at': '2026-01-11T13:00:00Z',
                  'sample': true,
                  'state': 'stale',
                },
                'market_value': '45',
                'unrealized_pnl': '14',
                'issue': null,
              },
            ]),
  'issues':
      issues ??
      (status == 'stale_sample'
          ? const [
              {'code': 'sample_data', 'message': 'local sample prices'},
            ]
          : const []),
};

Json unavailableValuationJson() => holdingValuationJson(
  status: 'unavailable',
  sample: true,
  totals: null,
  lines: const [
    {
      'symbol': 'AAPL',
      'quantity': '3',
      'cost_basis': '31',
      'currency': 'USD',
      'status': 'valued',
      'price': {
        'source': 'local_fixture',
        'venue': 'XNAS',
        'currency': 'USD',
        'price': '15',
        'price_adjustment': 'unspecified',
        'observed_at': '2026-01-10T14:00:00Z',
        'fetched_at': '2026-01-10T14:00:01Z',
        'recorded_at': '2026-01-11T13:00:00Z',
        'sample': true,
        'state': 'stale',
      },
      'market_value': '45',
      'unrealized_pnl': '14',
      'issue': null,
    },
    {
      'symbol': '005930',
      'quantity': '2',
      'cost_basis': '140000',
      'currency': 'KRW',
      'status': 'missing',
      'price': null,
      'market_value': null,
      'unrealized_pnl': null,
      'issue': {
        'code': 'missing_security_price',
        'message': 'eligible price unavailable',
        'field': '005930',
      },
    },
  ],
  issues: const [
    {
      'code': 'missing_security_price',
      'message': 'eligible price unavailable',
      'field': '005930',
    },
  ],
);

void main() {
  test('holding valuation GET has no query and parses exact strings', () async {
    late http.Request request;
    final api = RestOmniApi(
      baseUrl: 'http://example.test/',
      client: MockClient((value) async {
        request = value;
        return http.Response(jsonEncode(holdingValuationJson()), 200);
      }),
    );

    final valuation = await api.holdingValuation();

    expect(request.method, 'GET');
    expect(request.url.path, '/v1/portfolio/holding-valuation');
    expect(request.url.hasQuery, isFalse);
    expect(valuation.lines.single.marketValue, '45');
    expect(valuation.totals.single.unrealizedPnl, '14');
    expect(valuation.maxObservationAgeSeconds, 86400);
  });

  test('holding valuation parser rejects field and state drift', () {
    final valid = holdingValuationJson();
    final validLine = Json.from((valid['lines'] as List).single as Json);
    final validPrice = Json.from(validLine['price'] as Json);

    for (final invalid in <Json>[
      {...valid, 'account_id': 'private'},
      {
        ...valid,
        'lines': [
          {...validLine, 'instrument_id': 'private'},
        ],
      },
      {
        ...valid,
        'lines': [
          {
            ...validLine,
            'price': {...validPrice, 'observation_id': 'private'},
          },
        ],
      },
      {...valid, 'status': 'complete'},
      {...valid, 'sample': false},
      {...valid, 'max_observation_age_seconds': 86400.0},
      {...valid, 'totals': null},
      {
        ...valid,
        'lines': [
          {...validLine, 'market_value': 45},
        ],
      },
      {
        ...valid,
        'lines': [
          {...validLine, 'market_value': '0'},
        ],
      },
      {
        ...valid,
        'lines': [
          {...validLine, 'market_value': '-1'},
        ],
      },
      {
        ...valid,
        'lines': [
          {
            ...validLine,
            'price': {...validPrice, 'source': 'kiwoom_production'},
          },
        ],
      },
      {
        ...valid,
        'lines': [
          {...validLine, 'status': 'missing'},
        ],
      },
    ]) {
      expect(
        () => HoldingValuation.fromJson(invalid),
        throwsFormatException,
        reason: '$invalid',
      );
    }

    expect(
      HoldingValuation.fromJson(unavailableValuationJson()).totals,
      isEmpty,
    );
    expect(
      HoldingValuation.fromJson(
        holdingValuationJson(
          status: 'empty',
          sample: false,
          totals: null,
          lines: const [],
          issues: const [],
        ),
      ).lines,
      isEmpty,
    );

    final oneNanosecondStalePrice = {
      ...validPrice,
      'observed_at': '2026-01-10T13:59:59.999999999Z',
      'fetched_at': '2026-01-10T14:00:00Z',
    };
    final staleIssue = const {
      'code': 'stale_security_price',
      'message': 'price is older than policy',
      'field': 'AAPL',
    };
    expect(
      HoldingValuation.fromJson(
        holdingValuationJson(
          status: 'unavailable',
          sample: true,
          lines: [
            {
              ...validLine,
              'status': 'stale',
              'price': oneNanosecondStalePrice,
              'market_value': null,
              'unrealized_pnl': null,
              'issue': staleIssue,
            },
          ],
          issues: [staleIssue],
        ),
      ).lines.single.status,
      'stale',
    );
    expect(
      () => HoldingValuation.fromJson({
        ...valid,
        'lines': [
          {...validLine, 'price': oneNanosecondStalePrice},
        ],
      }),
      throwsFormatException,
    );
    expect(
      () => HoldingValuation.fromJson({
        ...valid,
        'lines': [
          {
            ...validLine,
            'price': {
              ...validPrice,
              'observed_at': '2026-01-10T14:00:00.000000002Z',
              'fetched_at': '2026-01-10T14:00:00.000000001Z',
            },
          },
        ],
      }),
      throwsFormatException,
    );
  });

  testWidgets('Holdings header opens the independent stored valuation', (
    tester,
  ) async {
    final api = _TestApi();
    await tester.pumpWidget(OmniFolioApp(api: api));
    await tester.pumpAndSettle();
    await tester.tap(find.text('보유'));
    await tester.pumpAndSettle();

    expect(find.text('저장 가격으로 평가 보기'), findsOneWidget);
    await tester.tap(find.text('저장 가격으로 평가 보기'));
    await tester.pumpAndSettle();

    expect(find.text('저장 가격 평가'), findsOneWidget);
    expect(find.text('AAPL'), findsOneWidget);
    expect(api.valuationCalls, 1);
  });

  testWidgets('empty Holdings still opens the independent stored valuation', (
    tester,
  ) async {
    final api = _TestApi(
      snapshotValue: PortfolioSnapshot(
        ledgerRevision: 'rev_0000000001',
        costBasisPolicy: 'fifo_exact_else_half_even_residual_8_v1',
        recordedAt: DateTime.utc(2026, 1, 11, 13),
        cash: const [],
        holdings: const [],
        realizedPnl: const [],
      ),
    );
    await tester.pumpWidget(OmniFolioApp(api: api));
    await tester.pumpAndSettle();
    await tester.tap(find.text('보유'));
    await tester.pumpAndSettle();

    expect(find.text('저장 가격으로 평가 보기'), findsOneWidget);
    await tester.tap(find.text('저장 가격으로 평가 보기'));
    await tester.pumpAndSettle();
    expect(api.valuationCalls, 1);
  });

  testWidgets(
    'sample and unavailable views explain authority and missing data',
    (tester) async {
      final api = _TestApi();
      await tester.pumpWidget(
        MaterialApp(home: HoldingValuationPage(api: api)),
      );
      await tester.pumpAndSettle();

      expect(find.text('샘플 저장 가격 · 실시간/현재 시세 아님'), findsOneWidget);
      expect(find.text('통화별 합계'), findsOneWidget);
      expect(find.text('미실현 이익 USD +14'), findsAtLeastNWidgets(1));
      expect(find.textContaining('최대 24시간'), findsOneWidget);

      api.valuationHandler = () async =>
          HoldingValuation.fromJson(unavailableValuationJson());
      await tester.tap(find.byTooltip('저장 가격 평가 새로고침'));
      await tester.pumpAndSettle();

      expect(find.text('일부 종목을 평가할 수 없어 합계를 표시하지 않습니다.'), findsOneWidget);
      expect(find.text('통화별 합계'), findsNothing);

      await tester.scrollUntilVisible(
        find.text('005930'),
        300,
        scrollable: find.descendant(
          of: find.byKey(const Key('holding-valuation-list')),
          matching: find.byType(Scrollable),
        ),
      );
      await tester.pumpAndSettle();

      expect(find.text('저장된 적격 가격이 없습니다.'), findsOneWidget);
      expect(find.text('005930'), findsOneWidget);
    },
  );

  testWidgets('refresh retains known-good data and redacts thrown details', (
    tester,
  ) async {
    final api = _TestApi();
    await tester.pumpWidget(MaterialApp(home: HoldingValuationPage(api: api)));
    await tester.pumpAndSettle();
    expect(find.text('AAPL'), findsOneWidget);

    api.valuationHandler = () async =>
        throw const ApiException('account-secret observation-secret');
    await tester.tap(find.byTooltip('저장 가격 평가 새로고침'));
    await tester.pumpAndSettle();

    expect(find.text('AAPL'), findsOneWidget);
    expect(find.text('새로고침하지 못했습니다. 이전 평가 결과를 유지합니다.'), findsOneWidget);
    expect(find.textContaining('account-secret'), findsNothing);
    expect(find.textContaining('observation-secret'), findsNothing);
  });

  testWidgets('empty known-good retains scope and refresh failure', (
    tester,
  ) async {
    final api = _TestApi()
      ..valuationHandler = () async => HoldingValuation.fromJson(
        holdingValuationJson(
          status: 'empty',
          sample: false,
          totals: null,
          lines: const [],
          issues: const [],
        ),
      );
    await tester.pumpWidget(MaterialApp(home: HoldingValuationPage(api: api)));
    await tester.pumpAndSettle();

    expect(find.textContaining('평가할 보유 종목이 없습니다'), findsOneWidget);
    expect(find.textContaining('현금 제외 · 전체 계좌 평가 아님'), findsOneWidget);
    expect(find.textContaining('rev_0000000001'), findsOneWidget);

    api.valuationHandler = () async => throw StateError('private-empty-error');
    await tester.tap(find.byTooltip('저장 가격 평가 새로고침'));
    await tester.pumpAndSettle();

    expect(find.text('새로고침하지 못했습니다. 이전 평가 결과를 유지합니다.'), findsOneWidget);
    expect(find.textContaining('평가할 보유 종목이 없습니다'), findsOneWidget);
    expect(find.textContaining('private-empty-error'), findsNothing);
  });

  testWidgets('retry is single-flight and an initial error is retryable', (
    tester,
  ) async {
    final api = _TestApi()
      ..valuationHandler = () async =>
          throw const ApiException('private-backend-detail');
    await tester.pumpWidget(MaterialApp(home: HoldingValuationPage(api: api)));
    await tester.pumpAndSettle();

    expect(find.text('저장 가격 평가를 불러오지 못했습니다'), findsOneWidget);
    expect(find.textContaining('private-backend-detail'), findsNothing);

    final retry = Completer<HoldingValuation>();
    api.valuationHandler = () => retry.future;
    await tester.tap(find.text('다시 시도'));
    await tester.tap(find.text('다시 시도'), warnIfMissed: false);
    await tester.pump();
    expect(api.valuationCalls, 2);
    retry.complete(HoldingValuation.fromJson(holdingValuationJson()));
    await tester.pumpAndSettle();
    expect(find.text('AAPL'), findsOneWidget);
  });

  testWidgets('320px at 200 percent keeps controls and semantic facts usable', (
    tester,
  ) async {
    tester.view.physicalSize = const Size(320, 760);
    tester.view.devicePixelRatio = 1;
    addTearDown(tester.view.resetPhysicalSize);
    addTearDown(tester.view.resetDevicePixelRatio);
    final semantics = tester.ensureSemantics();

    await tester.pumpWidget(
      MaterialApp(
        theme: ThemeData.light(),
        darkTheme: ThemeData.dark(),
        builder: (context, child) => MediaQuery(
          data: MediaQuery.of(
            context,
          ).copyWith(textScaler: const TextScaler.linear(2)),
          child: child!,
        ),
        home: HoldingValuationPage(api: _TestApi()),
      ),
    );
    await tester.pumpAndSettle();

    final pageContext = tester.element(find.byType(HoldingValuationPage));
    expect(MediaQuery.textScalerOf(pageContext).scale(10), 20);
    await tester.scrollUntilVisible(
      find.text('AAPL'),
      400,
      scrollable: find.descendant(
        of: find.byKey(const Key('holding-valuation-list')),
        matching: find.byType(Scrollable),
      ),
    );
    await tester.pumpAndSettle();

    expect(tester.takeException(), isNull);
    expect(find.byTooltip('저장 가격 평가 새로고침'), findsOneWidget);
    expect(
      find.semantics.byLabel(
        RegExp(r'AAPL.*수량 3.*원가 USD 31.*평가금액 USD 45.*미실현 이익 USD \+14'),
      ),
      findsOneWidget,
    );

    await tester.pumpWidget(
      MaterialApp(
        theme: ThemeData.light(),
        darkTheme: ThemeData.dark(),
        themeMode: ThemeMode.dark,
        home: HoldingValuationPage(api: _TestApi()),
      ),
    );
    await tester.pumpAndSettle();
    expect(
      Theme.of(tester.element(find.byType(HoldingValuationPage))).brightness,
      Brightness.dark,
    );
    expect(tester.takeException(), isNull);
    semantics.dispose();
  });
}

class _TestApi implements OmniApi {
  _TestApi({this.snapshotValue});

  int valuationCalls = 0;
  Future<HoldingValuation> Function()? valuationHandler;
  final PortfolioSnapshot? snapshotValue;

  @override
  Future<HoldingValuation> holdingValuation() async {
    valuationCalls += 1;
    return valuationHandler?.call() ??
        HoldingValuation.fromJson(holdingValuationJson());
  }

  @override
  Future<ServiceStatus> status() async => ServiceStatus.fromJson({
    'live_enabled': false,
    'mode': 'local_import_only',
    'trust_state': 'verified',
    'ledger_revision': 'rev_0000000001',
    'last_verified_at': '2026-01-11T13:00:00Z',
    'issues': const [],
  });

  @override
  Future<PortfolioSnapshot> snapshot() async =>
      snapshotValue ??
      PortfolioSnapshot(
        ledgerRevision: 'rev_0000000001',
        costBasisPolicy: 'fifo_exact_else_half_even_residual_8_v1',
        recordedAt: DateTime.utc(2026, 1, 11, 13),
        cash: const [],
        holdings: const [
          Holding(
            symbol: 'AAPL',
            quantity: '3',
            costBasis: '31',
            currency: 'USD',
          ),
        ],
        realizedPnl: const [],
      );

  @override
  Future<BrokerReconciliation?> latestBrokerReconciliation() async => null;

  @override
  Future<LocalOrderLog> localOrders() async => const LocalOrderLog(orders: []);

  @override
  Future<PaperMonitor> paperMonitor() =>
      throw UnsupportedError('Not used by this fixture.');

  @override
  Future<LedgerActivityPage> ledgerActivities() async =>
      const LedgerActivityPage(
        source: 'local_ledger',
        brokerFreshness: 'unverified',
        ledgerRevision: 'rev_0000000001',
        recordedAt: '2026-01-11T13:00:00Z',
        events: [],
        nextCursor: null,
      );

  @override
  Future<MarketCandles> candles(String symbol) =>
      throw UnsupportedError('Not used.');

  @override
  Future<ImportPreview> preview(String csv) =>
      throw UnsupportedError('Not used.');

  @override
  Future<ApplyReceipt> apply(String previewId, String idempotencyKey) =>
      throw UnsupportedError('Not used.');
}
